subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/AbstractMutableModuleComponentResolveMetadata.java

```diff
@@ -321,12 +321,18 @@ public List<Capability> getCapabilities() {
 
         @Override
         public void addDependency(String group, String module, VersionConstraint versionConstraint, List<ExcludeMetadata> excludes, String reason, ImmutableAttributes attributes, List<? extends Capability> requestedCapabilities, boolean endorsing, @Nullable IvyArtifactName artifact) {
-            dependencies.add(new DependencyImpl(group, module, versionConstraint, excludes, reason, attributes, requestedCapabilities, endorsing, artifact));
+            DependencyImpl dependency = new DependencyImpl(group, module, versionConstraint, excludes, reason, attributes, requestedCapabilities, endorsing, artifact);
+            if (!dependencies.contains(dependency)) {
```

Comment:
Should we consider switching `dependencies` (and `dependencyConstraints`) to a `LinkedHashSet`? This is changing from amortized O(1) insert to O(n) search + insert.  IIUC, this may only show up as a problem in modules with "lots" of direct dependencies or constraints.

Reply:
Ended up moving the de-duplication logic and using a `LinkedHashSet` in a class where that collection is only needed in a temporary fashion.